### PR TITLE
add ML Model get API [In progress]

### PR DESCRIPTION
**Pull request by Zhangxunmt**
### Description
This change added the support for Get Model API for ml-common plugin.
 
### Issues Resolved
Add Model-Get API.
 
### Check List
This is the first PR for the change, which is still subject to refine.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: common/src/main/java/org/opensearch/ml/common/parameter/MLModelOutput.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+import lombok.Builder;
+import lombok.Getter;
+import org.opensearch.common.io.stream.StreamInput;
+import org.opensearch.common.io.stream.StreamOutput;
+import org.opensearch.common.xcontent.XContentBuilder;
+
+import java.io.IOException;
+
+@Getter
+public class MLModelOutput extends MLOutput {
```
Comment (ylwu-amzn) [inline common/src/main/java/org/opensearch/ml/common/parameter/MLModelOutput.java:12]: We only need to add new `MLOutput` for new algorithm. For this get model API, we don't introduce new algorithm. So no need to add this class.

File: common/src/main/java/org/opensearch/ml/common/parameter/MLOutputType.java [hunk 2]
```diff
@@ -16,5 +16,6 @@
     TRAINING,
     PREDICTION,
     SAMPLE_ALGO,
-    SAMPLE_CALCULATOR
+    SAMPLE_CALCULATOR,
+    MODEL
```
Comment (ylwu-amzn) [inline common/src/main/java/org/opensearch/ml/common/parameter/MLOutputType.java:20]: No need to add model type

File: common/src/main/java/org/opensearch/ml/common/parameter/MLModelOutput.java (hunk 1 above)
Comment (Zhangxunmt) [reply common/src/main/java/org/opensearch/ml/common/parameter/MLModelOutput.java:12]: Correct. I think we still need to include MLModel in the MLModelGetResponse. Based on the current structure, the Response is defined in the ml-common lib, which is a dependency to ml-plugin where the MLModel is defined. So MLModelGetResponse cannot find the MLMode. My thinking is that we can move MLModel into ml-common, so it needs some refactor.

Comment (ylwu-amzn) [reply common/src/main/java/org/opensearch/ml/common/parameter/MLModelOutput.java:12]: Sure, feel free to move `MLModel` to `commons` package.

File: common/src/main/java/org/opensearch/ml/common/parameter/MLModel.java [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+        this.content = content;
+        this.user = user;
+    }
+
+    public MLModel(FunctionName algorithm, Model model) {
+        this(model.getName(), algorithm, model.getVersion(), Base64.getEncoder().encodeToString(model.getContent()), null);
+    }
+
+    public MLModel(StreamInput input) throws IOException{
+        name = input.readOptionalString();
```
Comment (ylwu-amzn) [inline common/src/main/java/org/opensearch/ml/common/parameter/MLModel.java:50]: Keep the same order when read from `StreamInput` as write out `StreamOutput`. The `writeTo` method write `content` after `name`, so we should read `content` after `name`.

File: common/src/main/java/org/opensearch/ml/common/parameter/Model.java [hunk 4]
```diff
@@ -12,5 +12,5 @@
 public class Model {
     String name;
     int version;
-    byte[] content;
+    public byte[] content;
```
Comment (ylwu-amzn) [inline common/src/main/java/org/opensearch/ml/common/parameter/Model.java:15]: Why make `content` public?

File: common/src/main/java/org/opensearch/ml/common/transport/model/MLModelGetAction.java [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.ml.common.transport.model;
+
+import org.opensearch.action.ActionType;
+
+public class MLModelGetAction extends ActionType<MLModelGetResponse> {
+    public static final MLModelGetAction INSTANCE = new MLModelGetAction();
+    public static final String NAME = "cluster:admin/opensearch/ml/models";
```
Comment (ylwu-amzn) [inline common/src/main/java/org/opensearch/ml/common/transport/model/MLModelGetAction.java:12]: Append "get" to action name? "cluster:admin/opensearch/ml/models/get"
We will have other model actions lile `delete`

File: plugin/src/main/java/org/opensearch/ml/action/models/GetModelTransportAction.java [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+                } catch (Exception e) {
+                    log.error("Failed to parse ml model" + r.getId(), e);
+                    actionListener.onFailure(e);
+                }
+            } else {
+                actionListener.onResponse(null);
+            }
+        }, e -> {
+            if (e instanceof IndexNotFoundException) {
+                actionListener.onResponse(null);
```
Comment (ylwu-amzn) [inline plugin/src/main/java/org/opensearch/ml/action/models/GetModelTransportAction.java:77]: Throw `ResourceNotFoundException`?

File: plugin/src/main/java/org/opensearch/ml/action/models/GetModelTransportAction.java [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+import org.opensearch.ml.common.transport.model.MLModelGetResponse;
+import org.opensearch.tasks.Task;
+import org.opensearch.transport.TransportService;
+
+@Log4j2
+@FieldDefaults(makeFinal = true, level = AccessLevel.PRIVATE)
+public class GetModelTransportAction extends HandledTransportAction<ActionRequest, MLModelGetResponse> {
+
+    TransportService transportService;
+    private final Client client;
```
Comment (ylwu-amzn) [inline plugin/src/main/java/org/opensearch/ml/action/models/GetModelTransportAction.java:38]: As line 34 has define filed as private and final, I think we can remove `private final` here

File: plugin/src/main/java/org/opensearch/ml/plugin/MachineLearningPlugin.java [hunk 8]
```diff
@@ -48,11 +50,7 @@
 import org.opensearch.ml.engine.algorithms.sample.LocalSampleCalculator;
 import org.opensearch.ml.indices.MLIndicesHandler;
 import org.opensearch.ml.indices.MLInputDatasetHandler;
-import org.opensearch.ml.rest.RestMLExecuteAction;
-import org.opensearch.ml.rest.RestMLPredictionAction;
-import org.opensearch.ml.rest.RestMLTrainAndPredictAction;
-import org.opensearch.ml.rest.RestMLTrainingAction;
-import org.opensearch.ml.rest.RestStatsMLAction;
+import org.opensearch.ml.rest.*;
```
Comment (ylwu-amzn) [inline plugin/src/main/java/org/opensearch/ml/plugin/MachineLearningPlugin.java:53]: Avoid wildcard import https://www.jetbrains.com/help/idea/creating-and-optimizing-imports.html#disable-wildcard-imports

File: common/src/main/java/org/opensearch/ml/common/parameter/MLModel.java (hunk 3 above)
Comment (Zhangxunmt) [reply common/src/main/java/org/opensearch/ml/common/parameter/MLModel.java:50]: Yeah will do in the revision.

File: common/src/main/java/org/opensearch/ml/common/parameter/Model.java (hunk 4 above)
Comment (Zhangxunmt) [reply common/src/main/java/org/opensearch/ml/common/parameter/Model.java:15]: Some unit tests in the plugin package directly use mode.content and intelliji auto detects them and set it to public. Have updated the UTs and removed the public.

File: common/src/main/java/org/opensearch/ml/common/transport/model/MLModelGetAction.java (hunk 5 above)
Comment (Zhangxunmt) [reply common/src/main/java/org/opensearch/ml/common/transport/model/MLModelGetAction.java:12]: cool will do.

File: plugin/src/main/java/org/opensearch/ml/action/models/GetModelTransportAction.java (hunk 6 above)
Comment (Zhangxunmt) [reply plugin/src/main/java/org/opensearch/ml/action/models/GetModelTransportAction.java:77]: sure will add them.

File: plugin/src/main/java/org/opensearch/ml/plugin/MachineLearningPlugin.java (hunk 8 above)
Comment (Zhangxunmt) [reply plugin/src/main/java/org/opensearch/ml/plugin/MachineLearningPlugin.java:53]: It has to be Intelliji replacing them to * or gradle.spotless task. I find this is a common problem using intellij.

File: plugin/src/main/java/org/opensearch/ml/action/models/GetModelTransportAction.java (hunk 7 above)
Comment (Zhangxunmt) [reply plugin/src/main/java/org/opensearch/ml/action/models/GetModelTransportAction.java:38]: sure, will remove them.

File: common/src/main/java/org/opensearch/ml/common/parameter/MLModel.java [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+                    version = parser.intValue();
+                    break;
+                case USER:
+                    user = User.parse(parser);
+                    break;
+                case ALGORITHM:
+                    algorithm = FunctionName.valueOf(parser.text());
+                    break;
+                default:
+                    break;
```
Comment (ylwu-amzn) [inline common/src/main/java/org/opensearch/ml/common/parameter/MLModel.java:125]: Add `parser.skipChildren();` before line125 to skip any wrong field name. Refer to AD code https://github.com/opensearch-project/anomaly-detection/blob/main/src/main/java/org/opensearch/ad/model/AnomalyDetector.java#L552

File: common/src/main/java/org/opensearch/ml/common/transport/model/MLModelGetRequest.java [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+
+    public MLModelGetRequest(StreamInput in) throws IOException {
+        super(in);
+        this.modelId = in.readOptionalString();
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        super.writeTo(out);
+        out.writeOptionalString(this.modelId);
```
Comment (ylwu-amzn) [inline common/src/main/java/org/opensearch/ml/common/transport/model/MLModelGetRequest.java:47]: As `modelId` is mandatory , we can just use `out.writeString`, same for line 41

**Comment by ylwu-amzn** [comment]
We'd better put the revision in a new commit after the first round of review, so reviewers can just check the new changes easily.

**Pull request merged**